api/grpcserver/v2alpha1/transaction_test.go

```diff
@@ -474,3 +479,224 @@ func newTx(nonce uint64, recipient types.Address, signer *signing.EdSigner) *typ
+		require.Equal(t, spacemeshv2alpha1.Transaction_TRANSACTION_TYPE_SINGLE_SIG_SPAWN, txType)
+	})
+
+	t.Run("singlesig send", func(t *testing.T) {
+		t.Parallel()
+
+		signer, err := signing.NewEdSigner()
+		require.NoError(t, err)
+		tx := newTx(1, types.Address{}, signer)
+
+		contents, txType, err := toTxContents(tx.Raw)
+		require.NoError(t, err)
+		require.NotNil(t, contents.GetSend())
+		require.Nil(t, contents.GetSingleSigSpawn())
+		require.Equal(t, spacemeshv2alpha1.Transaction_TRANSACTION_TYPE_SINGLE_SIG_SEND, txType)
+	})
+
+	t.Run("multisig spawn", func(t *testing.T) {
+		t.Parallel()
+
+		var pubs []ed25519.PublicKey
+		pks := make([]ed25519.PrivateKey, 0, 3)
+		for i := 0; i < 3; i++ {
+			pub, pk, err := ed25519.GenerateKey(rand.New(rand.NewSource(time.Now().UnixNano())))
+			require.NoError(t, err)
+			pubs = append(pubs, pub)
+			pks = append(pks, pk)
+		}
+
+		var agg *multisig2.Aggregator
+		for i := 0; i < len(pks); i++ {
+			part := multisig2.SelfSpawn(uint8(i), pks[i], multisig.TemplateAddress, 1, pubs, types.Nonce(1))
+			if agg == nil {
+				agg = part
+			} else {
+				agg.Add(*part.Part(uint8(i)))
+			}
+		}
+		rawTx := agg.Raw()
+
+		contents, txType, err := toTxContents(rawTx)
+		require.NoError(t, err)
+		require.NotNil(t, contents.GetMultiSigSpawn())
+		require.Nil(t, contents.GetSend())
+		require.Nil(t, contents.GetSingleSigSpawn())
+		require.Nil(t, contents.GetVestingSpawn())
+		require.Nil(t, contents.GetVaultSpawn())
+		require.Nil(t, contents.GetDrainVault())
+		require.Equal(t, spacemeshv2alpha1.Transaction_TRANSACTION_TYPE_MULTI_SIG_SPAWN, txType)
+	})
+
+	t.Run("multisig send", func(t *testing.T) {
+		t.Parallel()
+
+		var pubs []ed25519.PublicKey
+		pks := make([]ed25519.PrivateKey, 0, 3)
+		for i := 0; i < 3; i++ {
+			pub, pk, err := ed25519.GenerateKey(rand.New(rand.NewSource(time.Now().UnixNano())))
+			require.NoError(t, err)
+			pubs = append(pubs, pub)
+			pks = append(pks, pk)
+		}
+
+		to := wallet.Address(pubs[0])
+
+		var agg *multisig2.Aggregator
+		for i := 0; i < len(pks); i++ {
+			part := multisig2.Spend(uint8(i), pks[i], multisig.TemplateAddress, to, 100, types.Nonce(1))
+			if agg == nil {
+				agg = part
+			} else {
+				agg.Add(*part.Part(uint8(i)))
+			}
+		}
+		rawTx := agg.Raw()
+
+		contents, txType, err := toTxContents(rawTx)
+		require.NoError(t, err)
+		require.NotNil(t, contents.GetSend())
+		require.Nil(t, contents.GetMultiSigSpawn())
+		require.Nil(t, contents.GetSingleSigSpawn())
+		require.Nil(t, contents.GetVestingSpawn())
+		require.Nil(t, contents.GetVaultSpawn())
+		require.Nil(t, contents.GetDrainVault())
+		require.Equal(t, spacemeshv2alpha1.Transaction_TRANSACTION_TYPE_MULTI_SIG_SEND, txType)
+	})
+
+	t.Run("vault spawn", func(t *testing.T) {
+		t.Parallel()
+
+		var pubs []ed25519.PublicKey
+		pks := make([]ed25519.PrivateKey, 0, 3)
+		for i := 0; i < 3; i++ {
+			pub, pk, err := ed25519.GenerateKey(rand.New(rand.NewSource(time.Now().UnixNano())))
```

Comment:
This might end up with duplicated keys on Windows, where time resolution is low (the rand will be seeded with same value multiple times).